nasl/nasl_text_utils.c

```diff
@@ -814,24 +828,63 @@ nasl_eregmatch (lex_ctxt *lexic)
                    pattern);
       return NULL;
     }
+  retc = alloc_typed_cell (DYN_ARRAY);
+  retc->x.ref_val = a = g_malloc0 (sizeof (nasl_array));
 
-  if (regexec (&re, string, (size_t) NS, subs, 0) != 0)
+  if (find_all == 0)
     {
-      regfree (&re);
-      return NULL;
+      if (regexec (&re, string, (size_t) NS, subs, 0) != 0)
+        {
+          regfree (&re);
+          return NULL;
+        }
+
+      int i;
+      for (i = 0; i < NS; i++)
+        if (subs[i].rm_so != -1)
+          {
+            v.var_type = VAR2_DATA;
+            v.v.v_str.s_siz = subs[i].rm_eo - subs[i].rm_so;
+            v.v.v_str.s_val = (unsigned char *) string + subs[i].rm_so;
+            (void) add_var_to_list (a, i, &v);
```

Comment:
Is this even make sense now? Why put that information in the `i`th place of `a`, if we only match once?
If `find_all` is true, it makes sense to use the `index` that will be incremented after every match, but this `i` is iterating through the `subs` not through the array, right?

Reply:
This is for backward compatibility. So the behavior is the same as before, when the option is FALSE, or it is not passed.